### PR TITLE
Add Prometheus Support for Choreo Connect

### PR DESCRIPTION
**Pull request by Avishka-Shamendra**
## Purpose
- Add Prometheus Support for Choreo Connect
    - [Internal Issue](https://github.com/wso2-enterprise/wso2-apim-internal/issues/4153)
    - [Public Issue](https://github.com/wso2/product-microgateway/issues/3427)

## Approach
- Expose metrics endpoints of router, enforcer and adapter to Prometheus.

## Documentation
- [Enabling prometheus metrics on CC](https://github.com/wso2/product-microgateway/issues/3427#issuecomment-1697142545)

## Related PRs
- Support PR (merged) - https://github.com/wso2-support/product-microgateway/pull/390

### CODE REVIEW

File: adapter/pkg/metrics/metrics.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+		avg, err := load.Avg()
+		if handleError(err, "Failed to read cpu load averages") {
+			return
+		}
+		loadAvg.WithLabelValues("1m").Set(avg.Load1)
+		loadAvg.WithLabelValues("5m").Set(avg.Load5)
+		loadAvg.WithLabelValues("15m").Set(avg.Load15)
+
+		// Sleep 5s before the next measurement
+		time.Sleep(5 * time.Second)
```
Comment (NomadXD) [inline adapter/pkg/metrics/metrics.go:155]: Shouldn't we add this as a configuration ? IMO metric collection interval should be configurable.

File: adapter/pkg/metrics/metrics.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+			ErrorCode: 1109,
+		})
+		return true
+	}
+	return false
+}
+
+// StartPrometheusMetricsServer initializes and starts the metrics server to expose metrics to prometheus.
+func StartPrometheusMetricsServer(port int32) {
+	go recordMetrics()
```
Comment (NomadXD) [inline adapter/pkg/metrics/metrics.go:174]: If this go routine exits due to an error or some reason, metrics collection won't happen even though the metrics server runs. So it's better to watch for any exit signals, log it and restart the metrics collection.

File: adapter/pkg/metrics/metrics.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+	"github.com/shirou/gopsutil/v3/cpu"
+	"github.com/shirou/gopsutil/v3/host"
+	"github.com/shirou/gopsutil/v3/load"
+	"github.com/shirou/gopsutil/v3/mem"
+	logger "github.com/wso2/product-microgateway/adapter/internal/loggers"
+	"github.com/wso2/product-microgateway/adapter/pkg/logging"
+)
+
+// PrometheusMetricType prometheus metric type
+const PrometheusMetricType = "prometheus"
```
Comment (slahirucd7) [inline adapter/pkg/metrics/metrics.go:42]: Shall we move this to the types.go ?

File: enforcer-parent/enforcer/src/main/resources/prometheus-jmx-config.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
+# --------------------------------------------------------------------
+# Copyright (c) 2023, WSO2 Inc. (http://wso2.com).
```
Comment (slahirucd7) [inline enforcer-parent/enforcer/src/main/resources/prometheus-jmx-config.yml:2]: Copyright (c) 2023, WSO2 LLC.

It's better if we can update other places in the same PR.

File: integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/setup/standalone/CcWithPrometheusMetricsEnabled.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                .withNewConfig("prometheus-metrics-enabled.toml")
+                .withVolumeMountDir("metricConfigs")
+                .build();
+        ccInstance.start();
+        ApictlUtils.addEnv("test");
+        ApictlUtils.login("test");
+        TimeUnit.SECONDS.sleep(5);
+
+    }
+
```
Comment (slahirucd7) [inline integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/setup/standalone/CcWithPrometheusMetricsEnabled.java:44]: Are we expecting add a test case here ?
Seems this file is not having a test case....

File: integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/metrics/PrometheusMetricEndpointsTestCase.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    public void testAdapterMetricEndpoint() throws Exception {
+        HttpResponse response = HttpClientRequest.doGet("http://localhost:19085/metrics");
+        Assert.assertNotNull(response);
+        int responseCode = response.getResponseCode();
+        Assert.assertEquals(200, responseCode, "Response code mismatched");
+        Assert.assertTrue(response.getData().contains("go_"), "Metric information not found in expected format");
+    }
+
+
+    @Test(description = "Test whether router metric endpoint is up")
```
Comment (slahirucd7) [inline integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/metrics/PrometheusMetricEndpointsTestCase.java:45]: Shall we change the test description. All test cases are having the same

File: adapter/internal/adapter/adapter.go [hunk 7]
```diff
@@ -196,6 +197,13 @@ func Run(conf *config.Config) {
 		})
 	}
 
+	// Start the metrics server
+	if conf.Adapter.Metrics.Enabled && conf.Adapter.Metrics.Type == metrics.PrometheusMetricType {
```
Comment (slahirucd7) [inline adapter/internal/adapter/adapter.go:201]: Isn't it better to consider ignoring the case here ?
`prometheus` , `Prometheus` , `PROMETHEUS`

File: adapter/pkg/metrics/metrics.go (hunk 1 above)
Comment (Avishka-Shamendra) [reply adapter/pkg/metrics/metrics.go]: Fixed [3d0c00a](https://github.com/wso2/product-microgateway/pull/3428/commits/3d0c00ae2550e2e70d96f1671a62f9103054caeb)

File: adapter/pkg/metrics/metrics.go (hunk 2 above)
Comment (Avishka-Shamendra) [reply adapter/pkg/metrics/metrics.go]: Fixed [3d0c00a](https://github.com/wso2/product-microgateway/pull/3428/commits/3d0c00ae2550e2e70d96f1671a62f9103054caeb)

File: integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/setup/standalone/CcWithPrometheusMetricsEnabled.java (hunk 5 above)
Comment (Avishka-Shamendra) [reply integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/setup/standalone/CcWithPrometheusMetricsEnabled.java:43]: This is just the setup class, tests are in PrometheusMetricEndpointsTestCase class

File: adapter/internal/adapter/adapter.go (hunk 7 above)
Comment (Avishka-Shamendra) [reply adapter/internal/adapter/adapter.go]: Fix [de37947](https://github.com/wso2/product-microgateway/pull/3428/commits/de379477a8a20379d29d4fa4a4e0bc48d3e0fb18)

File: integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/metrics/PrometheusMetricEndpointsTestCase.java (hunk 6 above)
Comment (Avishka-Shamendra) [reply integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/standalone/metrics/PrometheusMetricEndpointsTestCase.java]: Fix [de37947](https://github.com/wso2/product-microgateway/pull/3428/commits/de379477a8a20379d29d4fa4a4e0bc48d3e0fb18)
The descriptions now are,
- Test whether Enforcer metric endpoint is up
- Test whether Adapter metric endpoint is up
- Test whether Router metric endpoint is up

File: enforcer-parent/enforcer/src/main/resources/prometheus-jmx-config.yml (hunk 4 above)
Comment (Avishka-Shamendra) [reply enforcer-parent/enforcer/src/main/resources/prometheus-jmx-config.yml]: Fix [de37947](https://github.com/wso2/product-microgateway/pull/3428/commits/de379477a8a20379d29d4fa4a4e0bc48d3e0fb18)

File: adapter/pkg/metrics/metrics.go (hunk 3 above)
Comment (Avishka-Shamendra) [reply adapter/pkg/metrics/metrics.go]: Fix fbcc13e

File: adapter/internal/adapter/adapter.go [hunk 8]
```diff
@@ -196,6 +197,13 @@ func Run(conf *config.Config) {
...
 	}
 
+	// Start the metrics server
+	if conf.Adapter.Metrics.Enabled && strings.ToLower(conf.Adapter.Metrics.Type) == metrics.PrometheusMetricType {
+		logger.LoggerMgw.Info("Starting Prometheus Metrics Server ....")
+		go metrics.StartPrometheusMetricsServer(conf.Adapter.Metrics.Port, conf.Adapter.Metrics.CollectionInterval)
+
+	}
+
 	logger.LoggerMgw.Info("Starting adapter ....")
```
Comment (slahirucd7) [inline adapter/internal/adapter/adapter.go:207]: Can we keep logs in this format ?

```suggestion
logger.LoggerMgw.Info("Starting adapter ....")
	
	// Start the metrics server
	if conf.Adapter.Metrics.Enabled && strings.ToLower(conf.Adapter.Metrics.Type) == metrics.PrometheusMetricType {
		logger.LoggerMgw.Info("Starting Prometheus Metrics Server ....")
		go metrics.StartPrometheusMetricsServer(conf.Adapter.Metrics.Port, conf.Adapter.Metrics.CollectionInterval)

	}


```

**Review by slahirucd7 — DISMISSED** [review]

File: adapter/pkg/metrics/metrics.go [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+		}
+		freePhysicalMemory.Set(float64(v.Free))
+		usedVirtualMemory.Set(float64(v.Used))
+		totalVirtualMemory.Set(float64(v.Total))
+
+		percentages, err := cpu.Percent(0, false)
+		if handleError(err, "Failed to read cpu usage metrics") || len(percentages) == 0 {
+			return
+		}
+		systemCPULoad.Set(percentages[0])
```
Comment (renuka-fernando) [inline adapter/pkg/metrics/metrics.go:122]: Do we only want the percentage of CPU 1?

File: adapter/internal/adapter/adapter.go [hunk 10]
```diff
@@ -196,7 +197,16 @@ func Run(conf *config.Config) {
 		})
 	}
 
-	logger.LoggerMgw.Info("Starting adapter ....")
+logger.LoggerMgw.Info("Starting adapter ....")
+	
+	// Start the metrics server
+	if conf.Adapter.Metrics.Enabled && strings.ToLower(conf.Adapter.Metrics.Type) == metrics.PrometheusMetricType {
```
[the remark this answers is not in the document]
Comment (Avishka-Shamendra) [reply adapter/internal/adapter/adapter.go]: Fixed in [7b1668e](https://github.com/wso2/product-microgateway/pull/3428/commits/7b1668e3349545785f7702713c6e908a0e79b89f)

File: adapter/pkg/metrics/metrics.go (hunk 9 above)
Comment (Avishka-Shamendra) [reply adapter/pkg/metrics/metrics.go]: Fixed in [7b1668e](https://github.com/wso2/product-microgateway/pull/3428/commits/7b1668e3349545785f7702713c6e908a0e79b89f)

File: adapter/pkg/metrics/metrics.go [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+}
+
+// StartPrometheusMetricsServer initializes and starts the metrics server to expose metrics to prometheus.
+func StartPrometheusMetricsServer(port int32, collectionInterval int32) {
+	done := make(chan struct{}) // Channel to indicate recordMetrics routine exit
+
+	// Start the Prometheus metrics server
+	go func() {
+		http.Handle("/metrics", promhttp.HandlerFor(prometheusMetricRegistry, promhttp.HandlerOpts{}))
+		err := http.ListenAndServe(":"+strconv.FormatInt(int64(port), 10), nil)
```
[the remark this answers is not in the document]
Comment (Avishka-Shamendra) [reply adapter/pkg/metrics/metrics.go]: Fixed in [7b1668e](https://github.com/wso2/product-microgateway/pull/3428/commits/7b1668e3349545785f7702713c6e908a0e79b89f)

**Review by renuka-fernando — APPROVED** [review]

**Review by dushaniw — APPROVED** [review]

File: adapter/internal/adapter/adapter.go [hunk 12]
```diff
@@ -196,7 +197,16 @@ func Run(conf *config.Config) {
 		})
 	}
 
-	logger.LoggerMgw.Info("Starting adapter ....")
+logger.LoggerMgw.Info("Starting adapter ....")
```
Comment (Lakith-Rambukkanage) [inline adapter/internal/adapter/adapter.go:200]: Proper indentation can be added

File: adapter/pkg/metrics/metrics.go [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+				Message:   fmt.Sprintln("Prometheus metrics server error:", err),
+				Severity:  logging.MAJOR,
+				ErrorCode: 1110,
+			})
+		}
+	}()
+
+	for {
+		// Start the recordMetrics goroutine
+		go func() {
```
Comment (Lakith-Rambukkanage) [inline adapter/pkg/metrics/metrics.go:194]: Better to add a doc comment explaining the pattern usage of `go` to create the goroutine and the `done <- struct{}{}` and `<-done` signal for parallel(to main thread) but sequential(to each call for `recordMetrics`) execution of `recordMetrics` function.

Comment (Avishka-Shamendra) [reply adapter/pkg/metrics/metrics.go:194]: Fixed through PR https://github.com/wso2/product-microgateway/pull/3498

**Pull request merged**